From 426a7397ddc4831a750d132cfd53953dd832bba2 Mon Sep 17 00:00:00 2001
From: kwliu <kwliu@nuvoton.com>
Date: Mon, 10 Dec 2018 11:31:40 +0800
Subject: [PATCH] support nuvoton gpiochip

---
 gpio.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/gpio.c b/gpio.c
index 33a9ffe..84deb06 100644
--- a/gpio.c
+++ b/gpio.c
@@ -154,10 +154,11 @@ int get_gpio_base()
 			continue;
 		}
 
-		char label[14];
-		if (fgets(label, 14, fd) != NULL)
+		char label[32];
+		if (fgets(label, 32, fd) != NULL)
 		{
-			if (strcmp(label, "1e780000.gpio") == 0)
+			if ((strcmp(label, "1e780000.gpio") == 0) ||
+				(strcmp(label, "/pinctrl@f0800000/gpio@f0010000") == 0))
 			{
 				is_bmc = TRUE;
 			}
-- 
2.17.1

